From ee7d9dd4953a45b1eb7fed807bd900932e40d646 Mon Sep 17 00:00:00 2001
From: letsfindaway <me@letsfindaway.de>
Date: Fri, 24 Jun 2022 07:01:07 +0200
Subject: [PATCH 1/2] feat: add containsPrefix string utility

- check if a string list contains a prefix to a given string
- fix typo in nextDigitizedName function name
---
 src/frameworks/UBStringUtils.cpp | 15 ++++++++++++++-
 src/frameworks/UBStringUtils.h   |  3 ++-
 2 files changed, 16 insertions(+), 2 deletions(-)

diff --git a/src/frameworks/UBStringUtils.cpp b/src/frameworks/UBStringUtils.cpp
index ce579d2d0..ac2a9395b 100644
--- a/src/frameworks/UBStringUtils.cpp
+++ b/src/frameworks/UBStringUtils.cpp
@@ -31,6 +31,19 @@
 
 #include "core/memcheck.h"
 
+bool UBStringUtils::containsPrefix(const QStringList &prefixes, const QString &string, Qt::CaseSensitivity cs)
+{
+    for (const QString& prefix : prefixes)
+    {
+        if (string.startsWith(prefix, cs))
+        {
+            return true;
+        }
+    }
+
+    return false;
+}
+
 QStringList UBStringUtils::sortByLastDigit(const QStringList& sourceList)
 {
     // we look for a set of digit after non digits and before a .
@@ -72,7 +85,7 @@ QStringList UBStringUtils::sortByLastDigit(const QStringList& sourceList)
 }
 
 
-QString UBStringUtils::netxDigitizedName(const QString& source)
+QString UBStringUtils::nextDigitizedName(const QString& source)
 {
 
     // we look for a set of digit after non digits and at the end
diff --git a/src/frameworks/UBStringUtils.h b/src/frameworks/UBStringUtils.h
index 5844fdb5f..5d61728ac 100644
--- a/src/frameworks/UBStringUtils.h
+++ b/src/frameworks/UBStringUtils.h
@@ -41,9 +41,10 @@ class UBStringUtils
         ~UBStringUtils() {}
 
     public:
+        static bool containsPrefix(const QStringList& prefixes, const QString& string, Qt::CaseSensitivity cs);
         static QStringList sortByLastDigit(const QStringList& source);
 
-        static QString netxDigitizedName(const QString& source);
+        static QString nextDigitizedName(const QString& source);
 
         static QString toCanonicalUuid(const QUuid& uuid);
 

From 630cd0a11712b1cc92885b470fd60a260d00e007 Mon Sep 17 00:00:00 2001
From: letsfindaway <me@letsfindaway.de>
Date: Fri, 24 Jun 2022 07:08:57 +0200
Subject: [PATCH 2/2] perf: improve performance of building font model

- remove duplicates from custom font list
- remove font list entries in chunks from model
---
 src/domain/UBGraphicsTextItemDelegate.cpp | 18 +++++++++++++++---
 src/gui/UBResources.cpp                   |  2 ++
 2 files changed, 17 insertions(+), 3 deletions(-)

diff --git a/src/domain/UBGraphicsTextItemDelegate.cpp b/src/domain/UBGraphicsTextItemDelegate.cpp
index 8851cd604..6ccb2b7f2 100644
--- a/src/domain/UBGraphicsTextItemDelegate.cpp
+++ b/src/domain/UBGraphicsTextItemDelegate.cpp
@@ -263,12 +263,24 @@ void UBGraphicsTextItemDelegate::customize(QFontDialog &fontDialog)
 
             QStringList customFontList =  UBResources::resources()->customFontList();
             int index = 0;
-            foreach (QString dialogFontName, dialogFontNames){
-                if (safeWebFontNames.contains(dialogFontName, Qt::CaseInsensitive) || customFontList.contains(dialogFontName, Qt::CaseSensitive))
+            int remove = 0;
+
+            for (const QString& dialogFontName : dialogFontNames)
+            {
+                if (UBStringUtils::containsPrefix(safeWebFontNames, dialogFontName, Qt::CaseInsensitive) ||
+                        UBStringUtils::containsPrefix(customFontList, dialogFontName, Qt::CaseSensitive))
+                {
+                    stringListModel->removeRows(index, remove);
+                    remove = 0;
                     index++;
+                }
                 else
-                    stringListModel->removeRow(index);
+                {
+                    ++remove;
+                }
             }
+
+            stringListModel->removeRows(index, remove);
         }
     }
     QList<QComboBox*> comboBoxes = fontDialog.findChildren<QComboBox*>();
diff --git a/src/gui/UBResources.cpp b/src/gui/UBResources.cpp
index 10af5395b..7e4b2ffcc 100644
--- a/src/gui/UBResources.cpp
+++ b/src/gui/UBResources.cpp
@@ -88,4 +88,6 @@ void UBResources::buildFontList()
         int fontId = QFontDatabase::addApplicationFont(fontFile);
         mCustomFontList << QFontDatabase::applicationFontFamilies(fontId);
     }
+
+    mCustomFontList.removeDuplicates();
 }
